rfcs/0000-const-ub.md

```diff
@@ -0,0 +1,88 @@
+- Feature Name: `const_ub`
+- Start Date: 2020-10-10
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Define UB during const evaluation to lead to an unspecified result for the affected CTFE query, but not otherwise infect the compilation process.
+
+# Motivation
+[motivation]: #motivation
+
+So far, nothing is specified about what happens when `unsafe` code leads to UB during CTFE.
+This is a major blocker for stabilizing `unsafe` operations in const-contexts.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+There are some values that Rust needs to compute at compile-time.
+This includes the initial value of a `const`/`static`, and array lengths (and more general, const generics).
+Computing these initial values is called compile-time function evaluation (CTFE).
+CTFE in Rust is very powerful and permits running almost arbitrary Rust code.
+This begs the question, what happens when there is `unsafe` code and it causes Undefined Behavior (UB)?
+
+The answer is that in this case, the final value that is currently being executed is arbitrary.
+For example, when UB arises while computing an array length, then the final array length can be any `usize`, or it can be (partially) uninitialized memory.
+No guarantees are made about this final value, and it can be different depending on host and target architecture, compiler flags, and more.
+However, UB will not otherwise adversely affect the currently running compiler; type-checking and lints and everything else will work correctly given whatever the result of the CTFE computation is.
+
+Note, however, that this means compile-time UB can later cause runtime UB when the program is actually executed:
+for example, if there is UB while computing the initial value of a `Vec<i32>`, the result might be a completely invalid vector that causes UB at runtime when used in the program.
+
+Sometimes, the compiler might be able to detect such problems and show an error or warning about CTFE computation having gone wrong (for example, the compiler might detect when the array length ends up being uninitialized).
+But other times, this might not be the case -- UB is not reliably detected during CTFE.
+This can change from compiler version to compiler version: CTFE code that causes UB could build fine with one compiler and fail to build with another.
+(This is in accordance with the general policy that unsound code is not subject to strict stability guarantees.)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+When UB arises as part of CTFE, the result of this evaluation is an unspecified constant.
+The compiler might be able to detect that UB occurred and raise an error or a warning, but this is not mandated, and absence of lints does not imply absence of UB.
```

Comment:
I think in this section we should explain the various mechanisms for detecting UB and their respective "detection rate" of UB.

1. Pointer dereferences
    * Dangling, null and *allocation* out of bounds pointers are reliably detected
    * Out of bounds pointers that are within the allocation but outside any valid object are not detected
    * Generally no guarantees are made for dereferencing arbitrary pointers with arbitrary types
2. Uninitialized memory
    * Values may be backed by an uninitialized value, any use of that value will cause an error
    * Uninitialized memory in the final value of a constant is reliably detected, but intermediate computations may do arbitrary things
3. Pointer comparisons
    * Pointers cannot be compared, this is prevented statically and dynamically.
    * The [`guaranteed_eq`](https://doc.rust-lang.org/std/primitive.pointer.html#method.guaranteed_eq) method on pointers allows fuzzy comparison of pointers, where a valid implementation is to always returns `false` and the code relying on the return value must still be valid if `false` were returned in all cases. It is only to be used for performance reasons.
4. Intrinsics
    * Compiler intrinsics often have soundness rules for their arguments. e.g. [`exact_div`](https://doc.rust-lang.org/std/intrinsics/fn.exact_div.html) for `exact_div(x, y)` results in UB where
        * `x % y != 0` or,
        * `y == 0` or,
        * `x == T::MIN && y == -1`
      but we reliably detect this.
    * All intrinsics must detect all the UB that they can cause.
    * Intrinsics that cannot have UB should be marked as safe
5. transmute/union access/pointer dereferences
    * It is entirely up to the user to ensure that the results of type punning derefs or actual transmutes result in valid values and does not depend on `repr(Rust)` types or other unspecified layouts.

Reply:
> Out of bounds pointers that are within the allocation but outside any valid object are not detected

I am not entirely sure what this even means, this is using C terminology.

> Values may be backed by an uninitialized value, any use of that value will cause an error

Here, "use" means strictly "actually performing an arithmetic or logic operation that this value as a input". A mere copy will not be detected (since we do not check validity).

Reply:
> I think in this section we should explain the various mechanisms for detecting UB and their respective "detection rate" of UB.

I have to say this list seems to be awfully geared around what happens to be the current implementation... in particular if we consider RFCs as applying to *all* Rust compilers, this seems unnecessarily specific.

I'd say something much shorter, mire like:
* UB caused by inrinsics is detected.
* Dereferencing dangling pointers is detected. (But misalignment is not!)
* All the other UB [listed here](https://doc.rust-lang.org/reference/behavior-considered-undefined.html) is not detected.

Reply:
> I am not entirely sure what this even means, this is using C terminology ("object").

I tried to say we don't detect references that point to uninitialized memory. Like the extra capacity in a vector could be accessed by using `set_len` wrongly and then indexing into the uninitialized parts.

Which kind of relates to your next point. Yes, I did mean actual operations, not just copying values.

> I have to say this list seems to be awfully geared around what happens to be the current implementation... in particular if we consider RFCs as applying to all Rust compilers, this seems unnecessarily specific.

Do we want other Rust compilers to be more permissive? Your list is only missing uninit memory in final values and basic operations and pointer comparisons. While I get that pointer comparisons aren't really UB, they just aren't possible, at least doing math or so with uninitialized memory/undefined values should be UB. Or is there a difference between intrinsics and basic ops?

Reply:
Ah right we could also mention "uninit memory occurring as an input to an arithmetic, logical or control-flow operation".

> While I get that pointer comparisons aren't really UB, they just aren't possible, at least doing math or so with uninitialized memory/undefined values should be UB. Or is there a difference between intrinsics and basic ops?

I'd like not to touch on unconst operations in this RFC; that is a more tricky subject that I planned to leave for the future.

Reply:
I am not sure about mandating that we detect "Uninit memory in final values"... this also needs very careful wording (uninit memory in a `union` or padding is okay!).

Reply:
It's a nice thing to have, but since the user can just generate the same value by transmuting some value at runtime, indeed not really a thing that we *need* per se. Lets leave it out.

> I'd like not to touch on unconst operations in this RFC; that is a more tricky subject that I planned to leave for the future.

yea, I realized that while writing it, thus the weird wording.

Reply:
So... with the parts in this thread that we agreed upon added to the RFC, I think it's ready (though ready for what I'm not sure, opening a zulip discussion about the next steps in the const roadmap, I think I still have some things that I said I'd do that I haven't done there)

Reply:
Read for submission as a proper RFC I hope?

Reply:
yea ,but shouldn't we do the const roadmap blogpost first?

Reply:
Should we? Not sure how the blog post would block even proposing the RFC.

Reply:
It doesn't block it, it's more... I was under the impression that the lang team kinda expects the next const thing to be given to them to be the roadmap that they'd sign off.

Reply:
This RFC was at least mentioned at the last meeting so I feel it is fine to be doing it in parallel. We'd still need this even after they signed off the road-map so I do not see a dependency either way here.

Reply:
That's true.